packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -48,18 +49,58 @@ function findAllHotswappableChanges(
   const hotswappableResources = new Array<HotswapOperation>();
   let foundNonHotswappableChange = false;
   stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const nonHotswappableResourceFound = isNonHotswappableResourceChange(change);
     const lambdaFunctionShortCircuitChange = isHotswappableLambdaFunctionChange(logicalId, change, assetParamsWithEnv);
-    if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+    const stepFunctionShortCircuitChange = isHotswappableStepFunctionChange(logicalId, change, assetParamsWithEnv);
+
+    const requiresFullDeployment = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (stepFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+                                   (nonHotswappableResourceFound === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT);
+
+    const irrelevant = (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (stepFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) &&
+                       (nonHotswappableResourceFound === ChangeHotswapImpact.IRRELEVANT);
+
+    if (requiresFullDeployment) {
       foundNonHotswappableChange = true;
-    } else if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) {
+    } else if (irrelevant) {
       // empty 'if' just for flow-aware typing to kick in...
     } else {
-      hotswappableResources.push(lambdaFunctionShortCircuitChange);
+      if (typeof lambdaFunctionShortCircuitChange !== 'string') {
+        hotswappableResources.push(lambdaFunctionShortCircuitChange);
+      }
+
+      if (typeof stepFunctionShortCircuitChange !== 'string') {
+        hotswappableResources.push(stepFunctionShortCircuitChange);
+      }
```

Comment:
This is ugly 😕. I really hope my above change talking about the algorithm will stop this from being required. These `typeof` checks should not be needed.